Add `--git-metadata` flag to `buf push`

doriable opened the pull request:
This PR adds the `--git-metadata` flag to `buf push`:

- this flag is a "meta-flag" that uses Git source control state to help set `--label` flag(s), `--source-control-url`, and `--create-default-label` to the HEAD branch of the default Git remote
- this flag is only compatible with checkouts of Git source repositories
- this flag does not allow you to set `--source-control-url`, `--create-default-label`, `--label`, `--tag`, `--branch`, or `--draft` labels alongside

This PR also changes the default visibility for `--create-visibility` to "private". This means that users are no longer required to specify the `--create-visibility` flag when calling `buf push --create` -- it will default to creating a private repository if one does not already exist.

Also, this PR includes a changelog entry for all the changes above, as well as the `--label`, `--create-default-label`, and `--source-control-url` flags.

doriable commented on private/buf/cmd/buf/command/push/push.go:
```diff
@@ -347,6 +388,306 @@ func validateLabelFlagValues(flags *flags) error {
+	container appext.Container,
+	flags *flags,
+) ([]bufmodule.UploadOption, error) {
+	input, err := bufcli.GetInputValue(container, flags.InputHashtag, ".")
+	if err != nil {
+		return nil, err
+	}
```
Attempting to figure out a graceful way of surfacing an error if this is not a checkout of a Git repository (e.g. a source ref).

doriable replied:
We should validate for unchecked references here, so we can also use this to handle a `128`.

pkwarren commented on private/buf/cmd/buf/command/push/push.go:
```diff
@@ -347,6 +388,306 @@ func validateLabelFlagValues(flags *flags) error {
+				currentHEADCommit,
+			)), nil
+		}
+		return bufmodule.UploadWithSourceControlURL(fmt.Sprintf(
```
Won't this break if someone is using a git remote instead of a https one? We should properly handle rewriting origins like `git@github.com:git/git.git` into appropriate URLs too.

Also I think we need to validate that the remote matches well known formats for GitHub/GitLab and if it doesn't match don't build a URL.

pkwarren commented on private/buf/cmd/buf/command/push/push.go:
```diff
@@ -347,6 +388,306 @@ func validateLabelFlagValues(flags *flags) error {
+			currentHEADCommit,
+		)), nil
+	}
+	// Otherwise we sort by alphabetical order and return the first URL
```
This is only getting the URL to the remote - not building the full URL to the commit like above. I also think we should allow the user some control over which origin is used instead of returning the first sorted one or origin.

pkwarren commented on private/buf/cmd/buf/command/push/push.go:
```diff
@@ -347,6 +388,306 @@ func validateLabelFlagValues(flags *flags) error {
+		); err != nil {
+			return nil, err
+		}
+		remoteToURL[remote] = strings.TrimSpace(buffer.String())
```
Just to check, I ran `git remote set-url origin ""` and it worked somehow. We might want to only add to the map if it is non-empty here.

doriable replied:
Yeah, that's actually a good point, sorry, missed this earlier! Updating.

pkwarren commented on private/buf/cmd/buf/command/push/push.go:
```diff
@@ -347,6 +388,306 @@ func validateLabelFlagValues(flags *flags) error {
+// sed -n '/HEAD branch/s/.*: //p'
+func getHEADBranchFromGitRemoteOutput(output []byte) (string, error) {
+	// First find where "HEAD branch"
+	index := bytes.Index(output, []byte("HEAD branch: "))
```
Instead of dealing in bytes, I think this would by much easier to read/maintain with a bufio.Scanner, strings.Cut, etc.

doriable replied:
Ah yes, it's a lot better that way!

pkwarren commented on private/buf/cmd/buf/command/push/push.go:
```diff
@@ -347,6 +388,306 @@ func validateLabelFlagValues(flags *flags) error {
+	if gitDefaultRemoteHEADBranch, ok := remoteToHEADBranch[gitDefaultRemote]; ok {
+		return gitDefaultRemoteHEADBranch, nil
+	}
+	// Otherwise we sort by alphabetical order and return the first URL
```
```suggestion
	// Otherwise we sort by alphabetical order and return the first origin's branch
```

pkwarren commented on private/buf/cmd/buf/command/push/push.go:
```diff
@@ -347,6 +388,306 @@ func validateLabelFlagValues(flags *flags) error {
+func validateGitMetadataFlags(flags *flags) error {
+	if flags.GitMetadata {
+		var usedFlags []string
+		if flags.SourceControlURL != "" {
```
Not blocking but I feel like if we couldn't detect a source code URL, we should allow the user to specify one (or fail push with an error). Otherwise they might not realize that git-metadata is failing to discover information.

Alternatively, we could warn if we fail to detect a URL or a HEAD branch (if --create is specified).

Additionally, labels should be additive - so we'll add labels by default for git metadata but it may be that users want to specify additional labels on commits and we should just take the intersection of them both.

saquibmian replied:
> if we couldn't detect a source code URL, we should allow the user to specify one

I think we should fail if we can't detect a source control URL or default branch.

> we should just take the intersection of them both

We discussed this and decided that, for now, it's one or the other. `--git-metadata` opines that Git is the source of truth for labels. If there are more labels to associate, associate them in Git and we'll push them. If you need something outside of that, you can't use `--git-metadata`.

doriable replied:
> Additionally, labels should be additive - so we'll add labels by default for git metadata but it may be that users want to specify additional labels on commits and we should just take the intersection of them both.

Yes, I had a conversation with folks this morning, and we're definitely allowing users to add `--label`s to this.

For now, we are disallowing `--source-control-url` and `--create-default-label`, but this may change.

pkwarren commented on private/pkg/git/git.go:
```diff
@@ -135,3 +139,22 @@ type ListFilesAndUnstagedFilesOptions struct {
+// GetDefaultBranch returns the repository's default branch name. It attempts to read it from the
+// `.git/refs/remotes/origin/HEAD` file, and expects it to be pointing to a branch also in the
+// `origin` remote.
+func GetDefaultBranch(gitDirPath string) (string, error) {
```
This appears unused?

doriable replied:
Oh! I pulled this in and no longer needed it, and didn't realise i didn't clear this file before committing. Yes, that is correct, will be deleting.

pkwarren reviewed (commented):
Overall looks good. I think we could benefit from some tests on the git logic (we could easily run `git init` in a temp dir and test various scenarios).

saquibmian commented on private/buf/cmd/buf/command/push/push.go:
```diff
@@ -132,6 +144,33 @@ func (f *flags) Bind(flagSet *pflag.FlagSet) {
+	--%s for each Git tag and branch for this Git commit
+	--%s to the Git default branch (e.g. main) - this is only in effect if --%s is also set
+
+The source control URL and default branch is based on the default Git remote, %q. If you do not have this remote, we take the URL for the first remote based on alphabetical ordering.
```
I've worked at places where `origin` was not the default remote. I would instead suggest that we take any remote, provided it's the only one. If there is more than one remote, require the user to specify one? I don't think we should arbitrarily pick one, though. Either there's one, or origin, or fail.

Question: how do fork builds work? If someone forks bufbuild/buf and sends a PR, and we have a GitHub action that runs `buf push --git-metadata` for the PR, what happens?

doriable replied:
Yeah, so we had a conversation about this, and we will be erroring if `origin` doesn't exist, and printing a warning if the checkout has multiple remotes.

saquibmian commented on private/buf/cmd/buf/command/push/push.go:
```diff
@@ -347,6 +388,306 @@ func validateLabelFlagValues(flags *flags) error {
+	if err != nil {
+		return nil, err
+	}
+	branch, err := getCurrentGitBranch(ctx, runner, input)
+	if err != nil {
+		return nil, err
+	}
```
Interesting, I would expect this to be the set of all branches that point to this commit, not just the current branch. Wouldn't that get us closer to what is in Git, and also pushes on other branches no-ops due to squashing? Otherwise a single git commit that is on multiple branches will not have a corresponding commit in the BSR with multiple branches.

I can see the value in being context-centric though -- this isn't `sync`, after all. What do we think?

doriable replied:
Just updating the comment here based on the conversations we've had external to this PR: it would be all branches that point directly to the commit at the time of the command.

saquibmian commented on private/buf/cmd/buf/command/push/push.go:
```diff
@@ -347,6 +388,306 @@ func validateLabelFlagValues(flags *flags) error {
+	// We prioritize the Git default remote, "origin", URL
+	if gitDefaultRemoteURL, ok := remoteToURL[gitDefaultRemote]; ok {
+		// First we need to trim the `.git` suffix if there is one.
+		gitDefaultRemoteURL = strings.TrimSuffix(gitDefaultRemoteURL, ".git")
```
I think we need to parse this URL. We will likely want to special-case based on hostname. This URL can be in a few different formats, see here: https://git-scm.com/docs/git-clone#_git_urls

saquibmian commented on private/buf/cmd/buf/command/push/push.go:
```diff
@@ -347,6 +388,306 @@ func validateLabelFlagValues(flags *flags) error {
+	if err != nil {
+		return nil, err
+	}
+	labels := append(tags, branch)
```
I was just about to say, this may be empty. Additionally, what would happen if CI runs on a commit that is in a detached HEAD such that `git branch --show-current` yields no results, and there are no tags associated with the commit? I suspect you will end up with no labels, which will effectively push to the default label?

pkwarren replied:
That seems dangerous as a default - perhaps we require that a branch is discovered?

doriable replied:
I don't think that's necessarily dangerous -- I think that the implication should be that if you have a detached `HEAD` checkout, you would be pushing to default label, and if that is a problem, then your default label should be protected by policy checks. It's unlikely that if your CI that does this for `buf push` doesn't also have similar workflows for other things -- therefore it must  be safe to operate against detached checkouts, so I think this is reasonable default behaviour. A warning for empty labels is reasonable though, I think.

doriable commented on private/buf/bufcli/flags_args.go:
```diff
@@ -148,8 +148,8 @@ func BindCreateVisibility(flagSet *pflag.FlagSet, addr *string, flagName string,
-		"",
-		fmt.Sprintf(`The repository's visibility setting, if created. Must be set if --%s is set. Must be one of %s`, createFlagName, stringutil.SliceToString(allVisibiltyStrings)),
+		privateVisibility,
+		fmt.Sprintf(`The repository's visibility setting, if created. Can only be set if --%s is set. Defaults to %s if not set with --%s. Must be one of %s`, createFlagName, createFlagName, privateVisibility, stringutil.SliceToString(allVisibiltyStrings)),
```
Ah yes, and I need to swap the two. Sorry, I rewrote this message a few times ^^". It should be:

```
The repository's visibility setting, if created. Can only be set if --create is set. Defaults to private if this flag is not set with --create. Must be one of [public,private] (default "private")
```

doriable commented on private/buf/cmd/buf/command/push/push.go:
```diff
@@ -132,6 +144,33 @@ func (f *flags) Bind(flagSet *pflag.FlagSet) {
+			createDefaultLabelFlagName,
+			tagFlagName,
+			branchFlagName,
+			draftFlagName,
```
Yeah, I went back and forth on this one. I think you are right, given the flags are hidden, we should not expose them, but still check for them.

unmultimedio commented on CHANGELOG.md:
```diff
@@ -2,7 +2,18 @@
 ## [Unreleased]
 
-- No changes yet.
+- Change the default visibiity of `--create-visibility` flag for `buf push` to `private`
```
```suggestion
- Change the default visibility of `--create-visibility` flag for `buf push` to `private`
```

doriable commented on private/buf/cmd/buf/command/push/push.go:
```diff
@@ -132,6 +144,33 @@ func (f *flags) Bind(flagSet *pflag.FlagSet) {
+		fmt.Sprintf(
+			`Uses the Git source control state to set flag values. If this flag is set, we will use the following values for your flags:
+
+	--%s to <git remote URL>/<repository name>/commit/<commit sha> (e.g. https://github.com/acme/weather/commit/ffac537e6cbbf934b08745a378932722df287a53)
```
> Technically, the URL is dependent on the VCS type (bitbucket is commits).

Yeah, that is handled in the code, I can specify that it's `<route>` in the template.

doriable commented on private/buf/cmd/buf/command/push/push.go:
```diff
@@ -347,6 +388,306 @@ func validateLabelFlagValues(flags *flags) error {
+		if err := runner.Run(
+			ctx,
+			gitCommand,
+			command.RunWithArgs("config", "--get", fmt.Sprintf("remote.%s.url", remote)),
```
Yeah -- this is something I went back and forth on. And these values may be different. You could have `git config --get origin.remote.url` return the SSH vs. HTTPS. This is related to @saquibmian's comment below and also your other comments on empty URL's and also the SSH URL. Let me collate the comments below and I'll propose something.

doriable replied:
Consolidating URL conversation here: so I think we should make a best effort on URL. The approach I took is:
- check that the URL has a valid scheme, if not, print a warning and set no source control URL
- if it has a valid scheme, parse the hostname. if the hostname is one of known hostnames, then we construct a `https` url for the user with the commit sha. otherwise we return just the URL to the source control repository, since there is no way to ascertain the route to the specific commit.

saquibmian commented on private/buf/cmd/buf/command/push/push.go:
```diff
@@ -347,6 +394,415 @@ func validateLabelFlagValues(flags *flags) error {
+	// control URL to the commit.
+	// url.Path already has a "/" prefix.
+	if strings.Contains(url.Hostname(), githubHostname) || strings.Contains(url.Hostname(), gitlabHostname) {
+		return fmt.Sprintf(defaultGitRemoteURLFormat, scheme, url.Hostname(), strings.TrimSuffix(url.Path, ".git"), commitSHA), nil
```
Same idea here, there may be a non-standard port here.

saquibmian commented on private/buf/cmd/buf/command/push/push.go:
```diff
@@ -347,6 +394,415 @@ func validateLabelFlagValues(flags *flags) error {
+		}
+		remoteToHEADBranch[remote] = branch
+	}
+	// We always use the Git remote, "origin", to check for the `HEAD` branch as the
+	// "default" branch.
```
Why fetch the others, then?

doriable commented on private/buf/cmd/buf/command/push/push.go:
```diff
@@ -347,6 +394,415 @@ func validateLabelFlagValues(flags *flags) error {
+	if !gitOriginRemoteFound {
+		return nil, fmt.Errorf("Git remote %q not found", gitOriginRemote)
+	}
+	if len(remotes) > 1 {
+		logger.Warn(
+			fmt.Sprintf("More than one Git remote found, %q will be used for Git metadata: %v", gitOriginRemote, remotes),
+		)
+	}
```
So the decision was that we wouldn't fail if there is more than one remote (basically exactly because of the forked upstreams situation -- you may get additional remotes from forks), but we do fail if the remote we care about, `origin` isn't present, and we basically only care about `origin`.

I think if we are sticking to this, then the only relevant remote is only ever `origin`, so we may not even need to return any here and simply use this to validate remotes instead.

saquibmian replied:
> may not even need to return any here and simply use this to validate remotes instead

Did you continue to return all remotes in the event that we revert this decision? If so, maybe a comment to that effect would be helpful -- a number of my comments were about this.

doriable replied:
It wasn't so much in case we revert so much as things were a bit in-flight as code was being written. I think if that's the UX we're going for, it's totally fine to have the code more closely reflect that :) So i'm onboard with making these changes.

doriable commented on private/buf/cmd/buf/command/push/push.go:
```diff
@@ -347,6 +394,415 @@ func validateLabelFlagValues(flags *flags) error {
+	}
+	// Docs: https://git-scm.com/docs/git-clone#_git_urls
+	// Validate URL scheme
+	if !slices.Contains([]string{"ssh", "git", "http", "https", "ftp", "ftps"}, url.Scheme) {
```
So I think the issue here is that the current checkout may be using `ssh`, even though `https` is also available, so as long as the scheme conforms to valid Git schemes, I think we should accept it. And then you make an attempt to point at a valid URL based on the hostname.

doriable commented on private/buf/cmd/buf/command/push/push.go:
```diff
@@ -347,6 +394,415 @@ func validateLabelFlagValues(flags *flags) error {
+	// use the Bitbucket route "commits" to construct a source control URL to the commit
+	// url.Path already has a "/" prefix.
+	if strings.Contains(url.Hostname(), bitBucketHostname) {
+		return fmt.Sprintf(bitBucketRemoteURLFormat, scheme, url.Hostname(), strings.TrimSuffix(url.Path, ".git"), commitSHA), nil
```
I'm not sure how the port factors in here ^^" We're making an attempt at constructing a `https` url based on the hostname, which drops the port...

But yes, I can add some unit tests for this.

saquibmian replied:
How do we know what the URL of the UI is? It could be on a port? With self-hosted it's hard to know.

bufdev commented on private/buf/cmd/buf/command/push/push.go:
```diff
@@ -347,6 +394,415 @@ func validateLabelFlagValues(flags *flags) error {
+	return modifiedFiles, nil
+}
+
+func getGitRemotes(
```
Same

bufdev replied:
Why are we returning multiple remotes? There's comments on this below as well

doriable commented on private/buf/cmd/buf/command/push/push.go:
```diff
@@ -347,6 +394,415 @@ func validateLabelFlagValues(flags *flags) error {
+		// with a git checkout.
+		return nil, fmt.Errorf("unable to check input %q, please ensure this is a Git repository checkout: %w", input, err)
+	}
+	if len(uncommittedFiles) > 0 {
```
This has always been a part of the product spec -- we do not allow users to push unchecked/uncommitted references. This makes sense, since we are tagging this with Git commit information through the VCS -- if there are uncommitted changes being pushed, that commit information would not make sense.

doriable commented on private/buf/cmd/buf/command/push/push.go:
```diff
@@ -347,6 +394,415 @@ func validateLabelFlagValues(flags *flags) error {
+	if len(labels) > 0 {
+		return bufmodule.UploadWithLabels(labels...), nil
+	}
+	logger.Warn("No tags or branch found for HEAD, contents will be pushed to the default label")
```
Based on today's conversation, we will now be erroring if no refs are found, so going to resolve this :)

doriable commented:
Resolved old conversations around URL parsing and erroring behaviours (based on today's conversation), refactored git commands for metadata to `pkg/git`, and added some tests around source control url parsing.

saquibmian commented on private/buf/cmd/buf/command/push/push.go:
```diff
@@ -347,6 +384,158 @@ func validateLabelFlagValues(flags *flags) error {
+		return nil, err
+	}
+	if len(refs) == 0 {
+		return nil, fmt.Errorf("not tags or branch found for HEAD, %s", gitCommitSha)
```
```suggestion
		return nil, fmt.Errorf("not tags or branches found for HEAD, %s", gitCommitSha)
```

saquibmian commented on private/buf/cmd/buf/command/push/push.go:
```diff
@@ -347,6 +384,158 @@ func validateLabelFlagValues(flags *flags) error {
+		return nil, err
+	}
+	if sourceControlURL == "" {
+		return nil, fmt.Errorf("no valid source control URL parsed for git commit sha: %s", gitCommitSha)
```
> unable to determine source control URL for this repository; only GitHub/GitLab/BitBucket are supported

doriable commented on CHANGELOG.md:
```diff
@@ -58,6 +58,20 @@
   `features.field_presence = LEGACY_REQUIRED` in Editions files. This new rule is not active by
   default in existing `v1` and `v1beta1` configurations, for backwards-compatibility reasons.
   Migrate your config to `v2` to use them.
+- Add `buf beta registry label archive` and `buf beta registry label unarchive` commands
```
Oh sure, yeah -- I just adjusted the changelog in the chang I'm making for those commands, but yeah.

unmultimedio replied:
Unresolving, this item was duped, we need to remove it from here too.

doriable commented on private/pkg/git/url.go:
```diff
@@ -0,0 +1,111 @@
+// It checks the hostname of repository remote URL -- if it contains a known hostname,
+// github, gitlab, and bitbucket, we construct a URL. Otherwise, we return an empty source
+// control URL.
+func GetSourceControlURL(
```
So I waffled a little on this one -- but since these functions are not encapsulated by an interface, I ended up not adding it to `git.go` because I didn't want to conflate the two styles... I can move them to `git.go` if they are confusing.

doriable commented on private/pkg/git/url.go:
```diff
@@ -0,0 +1,111 @@
+	rawRepositoryURL string,
+	gitCommitSha string,
+) (string, error) {
+	repositoryURL, err := url.Parse(rawRepositoryURL)
```
Yeah -- this is reasonable. I'll implement some custom handling for the scp-like URLs.

doriable commented on private/pkg/git/ur_test.go:
```diff
@@ -0,0 +1,62 @@
+	})
+	t.Run("ssh, github", func(t *testing.T) {
+		t.Parallel()
+		sourceControlURL, err := parseSourceControlURL("ssh://git@github.com/user/repo.git", gitCommitSha)
```
That is what they show on the UI, but you can also clone in this format. But either way, I can add handling for the scp-style URL and test it here.

doriable commented on private/pkg/git/url.go:
```diff
@@ -0,0 +1,111 @@
+// It checks the hostname of repository remote URL -- if it contains a known hostname,
+// github, gitlab, and bitbucket, we construct a URL. Otherwise, we return an empty source
+// control URL.
+func GetSourceControlURL(
```
I like this -- this also works well with the style of the this package is in, so let me move some stuff around and implement the scp-style URL handling logic alongside.

saquibmian commented on private/pkg/git/git.go:
```diff
@@ -135,3 +135,46 @@ type ListFilesAndUnstagedFilesOptions struct {
+	HEADBranch() string
+	// Kind of remote (e.g. GitHub, GitLab, BitBucket, etc.).
+	Kind() RemoteKind
+	// Hostname is the host name parsed from the remote URL. If the remote is not an unknown
```
```suggestion
	// Hostname is the host name parsed from the remote URL. If the remote is an unknown
```

saquibmian commented on private/pkg/git/remote.go:
```diff
@@ -0,0 +1,231 @@
+)
+
+var (
+	ErrRemoteNotFound = errors.New("not found")
```
```suggestion
	ErrRemoteNotFound = errors.New("git remote not found")
```
In case the error bubbles up unhandled

saquibmian commented on private/pkg/git/remote.go:
```diff
@@ -0,0 +1,231 @@
+	hostname       string
+	repositoryPath string
+	headBranch     string
+}
```
There should be a `newRemote` right after this, rather than constructing the remote in various places.

saquibmian commented on private/pkg/git/remote.go:
```diff
@@ -0,0 +1,231 @@
+}
+
+// getRemoteHEADBranch returns the HEAD branch based on the given remote and given
+// directory. Querying the remote for the HEAD branch requires the passing the
```
```suggestion
// directory. Querying the remote for the HEAD branch requires passing the
```

saquibmian reviewed (commented):
Looks good, just a couple stylistic things.

pkwarren commented on private/buf/bufcli/flags_args.go:
```diff
@@ -148,8 +148,8 @@ func BindCreateVisibility(flagSet *pflag.FlagSet, addr *string, flagName string,
-		"",
-		fmt.Sprintf(`The repository's visibility setting, if created. Must be set if --%s is set. Must be one of %s`, createFlagName, stringutil.SliceToString(allVisibiltyStrings)),
+		privateVisibility,
+		fmt.Sprintf(`The repository's visibility setting, if created. Can only be set if --%s is set. Defaults to %s if this flag is not set with --%s. Must be one of %s`, createFlagName, privateVisibility, createFlagName, stringutil.SliceToString(allVisibiltyStrings)),
```
```suggestion
		fmt.Sprintf(`The repository's visibility setting, if created. Can only be set with --%s. Must be one of %s`, createFlagName, stringutil.SliceToString(allVisibiltyStrings)),
```

This will replace:
```
The repository's visibility setting, if created. Can only be set if --create is set. Defaults to private if this flag is not set with --create. Must be one of [public,private] (default "private")
```
with:
```
The repository's visibility setting, if created. Can only be set with --create. Must be one of [public,private] (default "private")
```

pkwarren commented on private/buf/cmd/buf/command/push/push.go:
```diff
@@ -132,6 +142,29 @@ func (f *flags) Bind(flagSet *pflag.FlagSet) {
+	--%s for each Git tag and branch pointing to the currently checked out commit
+	--%s to the Git default branch (e.g. main) - this is only in effect if --%s is also set
+
+The source control URL and default branch is based on the Git remote, %q, and requires you to have this remote.
```
```suggestion
The source control URL and default branch is based on the required Git remote %q.
```

pkwarren commented on private/buf/cmd/buf/command/push/push.go:
```diff
@@ -132,6 +142,29 @@ func (f *flags) Bind(flagSet *pflag.FlagSet) {
+
+The source control URL and default branch is based on the Git remote, %q, and requires you to have this remote.
+This flag is only compatible with checkouts of Git source repositories.
+This flag does not allow you to set any of the following flags yourself: --%s, --%s.`,
```
Do we need to mention anything about how any additional `--label` parameters will be added in addition to the default labels detected from Git?

doriable replied:
Hmm, it might not be a bad idea to do so -- the more information the better. I'll add something here for that :)

pkwarren commented on private/buf/cmd/buf/command/push/push.go:
```diff
@@ -347,6 +386,155 @@ func validateLabelFlagValues(flags *flags) error {
+// If the remote hostname contains github (e.g. github.mycompany.com or github.com) or gitlab
+// (e.g. gitlab.mycompany.com or gitlab.com) then it uses the route /commit for the git
+// commit sha.
+func getGitMetadataSourceControlURLUploadOption(
```
I think it would be good to move the building of source control URLs to the git package in order to keep the logic in here purely around upload. Perhaps a `SourceControlURL(commitSHA string)` on the `git.Remote` interface?

doriable replied:
Hmm, yeah, I went back and forth a little bit on where specifically this logic should live, since it's kind of "business-y" (rather than "generically git"), but also, the `RemoteKind` type already breaks that a little bit. So given that, it might make sense to put there.

It is a little strange to have this function take `commitSHA`, but I think that is reasonable. I'll make it clear that it just accepts the `commitSHA` as a string but doesn't do any validation against that string.

doriable replied:
Actually, in this case, we wouldn't need to expose `RemoteKind` at all. I think I'm going to unexport the enum and keep it since it is useful for keeping the parsing logic clean (and independent from the initial parsing of the URL).

pkwarren commented on private/pkg/git/remote.go:
```diff
@@ -51,7 +67,7 @@ func (r *remote) Name() string {
 }
 
-func (r *remote) Kind() RemoteKind {
+func (r *remote) Kind() remoteKind {
```
We can remove this method now if we want.

bufdev commented on private/buf/cmd/buf/command/push/push.go:
```diff
@@ -347,6 +385,117 @@ func validateLabelFlagValues(flags *flags) error {
+		return nil, err
+	}
+	runner := command.NewRunner()
+	uncommittedFiles, err := git.CheckForUncommittedGitChanges(ctx, runner, input)
```
To what I can see, there is no verification here that `input` in in fact a directory, however it is expected by `git.CheckForUncommitedGitChanges` that `input` is a path to a local directory. This validation needs to be done, and should be obvious in the context of this function.

doriable replied:
Added validation here.
It is worth noting that `git.CheckForUncommittedGitChanges` would return an error that indicates if the provided input is not a valid directory and was captured in the error handling below, but it's safer to more explicit with the validation here.

bufdev commented on private/buf/cmd/buf/command/push/push.go:
```diff
@@ -347,6 +385,117 @@ func validateLabelFlagValues(flags *flags) error {
+	if len(uncommittedFiles) > 0 {
+		return nil, fmt.Errorf("uncommitted changes found in the following files: %v", uncommittedFiles)
+	}
+	originRemote, err := git.GetRemote(ctx, runner, app.EnvironMap(container), input, gitOriginRemote)
```
Same comment re: input validation.

bufdev replied:
Other functions in the `git` package (and others) take containers, this function should continue to follow this pattern by taking an `app.EnvContainer` instead of passing a `map[string]string` via `app.EnvironMap`.

bufdev commented on private/pkg/git/remote.go:
```diff
@@ -0,0 +1,282 @@
+)
+
+var (
+	ErrRemoteNotFound = errors.New("git remote not found")
```
No Godoc, and should be in `git.go` - `remote.go` should be limited to `remote` and supporting functions called by `remote` methods

bufdev commented on private/pkg/git/remote.go:
```diff
@@ -0,0 +1,282 @@
+// remoteKind is the kind of remote based on Git source (e.g. GitHub, GitLab, BitBucket, etc.)
+type remoteKind int
+
+const (
```
Global consts always come first, then global vars, then types.

bufdev commented on private/pkg/git/metadata.go:
```diff
@@ -0,0 +1,144 @@
+
+// CheckForUncommittedGitChanges checks if there are any uncommitted and/or unchecked
+// changes from git based on the given directory.
+func CheckForUncommittedGitChanges(
```
Why were these functions split out into a separate file than `git.go`, but `Remote` was not?

doriable replied:
Oh yeah, I put this in a separate file initially because these didn't come with an additional interface abstraction. But I've moved them to `git.go`, since it does conform with the old Go package styles.

doriable replied:
Oh yeah, I put this in a separate file initially because these didn't come with an additional interface abstraction. But I've moved them to `git.go`, since it does conform with the old Go package styles.

bufdev commented on private/buf/cmd/buf/command/push/push.go:
```diff
@@ -347,6 +386,120 @@ func validateLabelFlagValues(flags *flags) error {
+		return nil, err
+	}
+	// validate that input is a local directory
+	fileInfo, err := os.Stat(input)
```
This is not sufficient to validate the input type. You need to parse this in the same manner as we do in the format command, to validate the input would be parsed into a DirRef, and you additionally need to do this validation

bufdev commented on private/buf/cmd/buf/command/push/push.go:
```diff
@@ -347,6 +386,120 @@ func validateLabelFlagValues(flags *flags) error {
+		return nil, err
+	}
+	if len(uncommittedFiles) > 0 {
+		return nil, fmt.Errorf("uncommitted changes found in the following files: %v", uncommittedFiles)
```
You should specify to the user that git-metadata requires there to be no uncommitted changes

bufdev commented on private/buf/cmd/buf/command/push/push.go:
```diff
@@ -347,6 +386,120 @@ func validateLabelFlagValues(flags *flags) error {
+	}
+	sourceControlURL := originRemote.SourceControlURL(currentGitCommit)
+	if sourceControlURL == "" {
+		return nil, appcmd.NewInvalidArgumentError("unable to determine source control URL for this repository; only GitHub/GitLab/BitBucket are supported")
```
Why does it matter? This is a severe limitation, and excludes any github enterprise customer. We should be able to parse a source control url for any http/https/ssh endpoint

nicksnyder replied:
There are two potential questions here so will answer both.

**Why is it an error if we can't infer a source control url?**
If the user provided `--git-metadata` then they are expecting us to automatically set some values. If we can't set the values they are expecting, we should error. This means during setup, they will catch the problem early, and then later if, due to a code host migration, it breaks, they will know instantly.

**Why are we limited to GitHub/Bitbucket/GitLab?**
The specification for Remote.SourceControlURL() (which is in the interface documentation in this PR) is designed to work for enterprise customers as long as the name of their code host is in the remote name. If it isn't, we don't know what code host we are dealing with and therefore we don't know how to correctly construct the user facing url because the routes are different depending on the code host.

On GitHub and GitLab, routes look like `https://<host><repo-path>/commit/<commit-sha>`
and on BitBucket, routes look like `https://<host><repo-path>/commits/<commit-sha>`
(note `/commit/` vs `/commits`).

bufdev replied:
The code quite literally checks for if the hostname includes `github, gitlab, bitbucket`, and just with a `strings.Contains` on the hostname. This both requires hostnames to include github, for example (this isn't a hard requirement), and it would also accept things like `foogithubbar`, which doesn't feel like what this filter wants to do in the first place.

This limitation isn't acceptable for our enterprise customers. Additionally, my impression was that "source control URL" wasn't mean to link to a specific point-in-time on the repository, it was just meant to link to the repository itself.

We can discuss deferring this until after release, but I consider this a bug - limiting acceptable hostnames to `strings.Contains` on one of `github, gitlab, bitbucket` isn't acceptable

nicksnyder commented on private/pkg/git/git.go:
```diff
@@ -135,3 +152,161 @@ type ListFilesAndUnstagedFilesOptions struct {
+	// is an unknown kind, then this may be an empty string.
+	RepositoryPath() string
+	// SourceControlURL makes the best effort to construct a user-facing source control url
+	// given a commit sha string based on the remoteKind and available hostname and
```
Seems weird to me to reference a private implementation detail like `remoteKind` in a public interface.

nicksnyder commented on private/pkg/git/git.go:
```diff
@@ -135,3 +152,161 @@ type ListFilesAndUnstagedFilesOptions struct {
+	// then it uses the route /commits for the git commit sha.
+	//
+	// If the remote hostname contains github (e.g. github.mycompany.com or github.com) or gitlab
+	// (e.g. gitlab.mycompany.com or gitlab.com) then it uses the route /commit for the git
```
I think it would be clearer if you documented the full route returned, like:

`https://<hostname><repository-path>/commit/<git-commit-sha>`


(and same for bitbucket)

nicksnyder commented on private/pkg/git/remote.go:
```diff
@@ -0,0 +1,277 @@
+	githubHostname              = "github"
+	gitlabHostname              = "gitlab"
+	gitSuffix                   = ".git"
+	githubGitlabRemoteURLFormat = "https://%s%s/commit/%s"
```
is it really worth it to smash these two cases together rather than having a separate variable for each, even if the value is the same?
